docs/terminals.md

```diff
@@ -0,0 +1,291 @@
+# Terminals and Standard IO #
+
+*Note that the default configuration of `runc` (foreground, new terminal) is
+generally the best option for most users. This document exists to help explain
+what the purpose of the different modes is, and to try to steer users away from
+common mistakes and misunderstandings.*
+
+In general, most processes on Unix (and Unix-like) operating systems have 3
+standard file descriptors provided at the start, collectively referred to as
+"standard IO" (`stdio`):
+
+* `0`: standard-in (`stdin`), the input stream into the process
+* `1`: standard-out (`stdout`), the output stream from the process
+* `2`: standard-error (`stderr`), the error stream from the process
+
+When creating and running a container via `runc`, it is important to take care
+to structure the `stdio` the new container's process receives. In some ways
+containers are just regular processes, while in other ways they're an isolated
+sub-partition of your machine (in a similar sense to a VM). This means that the
+structure of IO is not as simple as with ordinary programs (which generally
+just use the file descriptors you give them).
+
+## Other File Descriptors ##
+
+Before we continue, it is important to note that processes can have more file
+descriptors than just `stdio`. By default in `runc` no other file descriptors
+will be passed to the spawned container process. If you wish to explicitly pass
+file descriptors to the container you have to use the `--preserve-fds` option.
+These ancillary file descriptors don't have any of the strange semantics
+discussed further in this document (those only apply to `stdio`) -- they are
+passed untouched by `runc`.
+
+It should be noted that `--preserve-fds` does not take individual file
+descriptors to preserve. Instead, it takes how many file descriptors (not
+including `stdio` or `LISTEN_FDS`) should be passed to the container. In the
+following example:
+
+```
+% runc run --preserve-fds 5 <container>
+```
+
+`runc` will pass the first `5` file descriptors (`3`, `4`, `5`, `6`, and `7` --
+assuming that `LISTEN_FDS` has not been configured) to the container.
+
+In addition to `--preserve-fds`, `LISTEN_FDS` file descriptors are passed
+automatically to allow for `systemd`-style socket activation. They always take
+up the first file descriptors after `stdio` (and `--preserve-fds` always follow).
+To disable this, simply unset `LISTEN_FDS`.
```

Comment:
1. what happens if `LISTEN_FDS` are set and `--preserve-fds`? An example like the above might help?
2. By default, isn't `LISTEN_FDS` unset anyways?

Reply:
> 1. What happens if LISTEN_FDS are set and --preserve-fds? An example like the above might help?

The first `LISTEN_FDS` file descriptors come first, then the `--preserve-fds` ones (the `--help` for `runc run` explains it a bit). I could try to add an example, but `LISTEN_FDS` is a bit difficult to given an example of without having a `systemd` service and all the rest of that fun stuff.

> 2. By default, isn't LISTEN_FDS unset anyways?

Yup, but it is important to mention (because it's quite related to `--preserve-fds`).

Reply:
That helps, thanks.